### PR TITLE
DQM TrackerMonitorTrack improvements

### PR DESCRIPTION
**Pull request by schneiml**
This PR adds a number of changes on top of PR#13240.
- Location and naming of the residual histograms for SiPixel is improved
- Cuts on pT and dxy significance are applied for all residuals. Such cuts were missing in the original SiStrip version. The histograms are now closer to the old SiPixel residuals histograms.
- The source code now compiles into two modules to keep the configrations of SiPixel and SiStrip independent. 
- Some other cleanup.

These changes affect the configuration options for the TrackMonitorTrack module. I updated the config files as far as I could follow; however, the new SiPixelMonitorTrackResiduals might be added in more places. Also, some files use options that were probably broken before, i did not touch these.

### CODE REVIEW

**Comment by mmusich** [comment]
please test

File: Alignment/OfflineValidation/src/TrackerValidationVariables.cc [hunk 1]
```diff
@@ -355,18 +369,3 @@ TrackerValidationVariables::fillTrackQuantities(const edm::Event& event,
...
-  
-  LogDebug("TrackerValidationVariables") << "trajColl->size(): " << trajCollectionHandle->size() ;
-
-  for (std::vector<Trajectory>::const_iterator it = trajCollectionHandle->begin(), itEnd = trajCollectionHandle->end(); 
-       it!=itEnd;
-       ++it) {
-    
-    fillHitQuantities(&(*it), v_avhitout);
-  }
-}
```
Comment (mmusich) [inline Alignment/OfflineValidation/src/TrackerValidationVariables.cc]: @schneiml why is thie method removed?

Comment (schneiml) [inline Alignment/OfflineValidation/src/TrackerValidationVariables.cc]: The comment in the header says it is only used in TrackerMonitorTrack, which now uses the other version. [DXR confirms there are no other callers (as of 7_6_X)](
http://cmslxr.fnal.gov/dxr/CMSSW/search?q=%2Bfunction-ref%3A%22TrackerValidationVariables%3A%3AfillHitQuantities%28const+edm%3A%3AEvent+%26%2C+std%3A%3Avector%3CAVHitStruct%3E+%26%29%22)

**Comment by mmusich** [comment]
mmh seems the tests are stuck...

**Comment by mmusich** [comment]
@schneiml by the way do you have also a backport to 80X? I couldn't find it...

**Comment by schneiml** [comment]
I only opened this one against 81X. Maybe it should be backported as well (technically there are no problems).

**Comment by mmusich** [comment]
@schneiml well, the bulk of the PR is related more to DQM than AlCa, but I guess @boudoul and @deguio might be interested to have it in 80X for this year's data taking as well (at least I would be, for alignment monitoring purposes)

**Comment by schneiml** [comment]
Ok, #13404 now targets 80X.

**Comment by mmusich** [comment]
@schneiml thanks!

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
+1 for https://github.com/schneiml/cmssw/commit/9bca63f06fd2543651d9920a7095acfc6ed414f4
changes in AlCa package (`Alignment/OfflineValidation`) are in line with code description:
   * remove unnecessary interface `fillHitQuantities` from  `Alignment/OfflineValidation/interface/TrackerValidationVariables.h`
   * added filter functionality in `fillTrackQuantities` method of same class, to allow cuts on pT and dxy significance. Filter is always passed when running the standalone Tracker Alignment offline validation suite.

@tlampen, @ghellwig @mschrode scream now if necessary

**Comment by boudoul** [comment]
@schneiml<https://github.com/schneiml> and @mmusich :
yes good to target 8X indeed , thanks

Le 19 févr. 2016 à 10:35, Marcel Schneider <notifications@github.com<mailto:notifications@github.com>> a écrit :


Ok, #13404<https://github.com/cms-sw/cmssw/pull/13404> now targets 80X.

**Comment by boudoul** [comment]
hi @deguio , could you please consider reviewing this PR (and the corresponding one in 80X #13404 ) ?
Thanks a lot

**Comment by deguio** [comment]
ciao,
sorry for being late. I am not sure I understand the need of templating the analyzer. is it only for compatibility reasons?
from what I see it was just easier to maintain the same analyzer and clone/configure it from the python.
F.

**Comment by schneiml** [comment]
Switching the behaviour using configuration would also be possible. The template approach has the advantage of keeping the SiStrip stuff intact (by being compatible) and keeping the two completely separated on the configuration side (to my knowledge, there is no other plugin that is involved in SiPixel and SiStrip DQM). 

Also, i think having separate names is easier to understand in the configuration, and the template approach is about as readable in the implementation (there is a bit of boilerplate for all the template decorations, though).

**Comment by deguio** [comment]
+1
even when cloning the module in the python you can keep the configurations completely decoupled and the sistrip stuff unthouched. anyway this seems to work fine.

File: DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc [hunk 2]
```diff
@@ -229,28 +222,41 @@ void MonitorTrackResiduals::analyze(const edm::Event& iEvent, const edm::EventSe
...
-      histos.x.normed->Fill(it.resXprime/it.resXprimeErr);
-    }
-    if(it.resYprimeErr != 0 && histos.y.base) {
-      histos.y.base->Fill(it.resYprime);
-      histos.y.normed->Fill(it.resYprime/it.resYprimeErr);
+  avalidator_.fillTrackQuantities(iEvent, iSetup, 
+                  // tell the validator to only look at good tracks
+                  [](const reco::Track& track) -> bool { 
+                    return track.pt() > 0.75
+                        && abs( track.dxy() ) > 5*track.dxyError();
```
Comment (mmusich) [inline DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc]: @schneiml am I wrong or you are going to look at very displaced tracks w.r.t to (0,0,0)? what is the rationale of this cut SIP > 5? I don't think tracks are going to be good, and in any case the cut will have phi-dependence induced by the beamspot position.

Comment (schneiml) [inline DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc]: @mmusich These cuts are inspired by what SiPixelMonitorTrackResiduals does, but simplified. I remember that there was a selection of an optimal vertex that we dropped here. 

Overall, I don't understand enough about the detector to tell whether these make sense, but I discussed them with @fioriNTU and we settled on these.

Comment (mmusich) [inline DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc]: @fioriNTU can you please comment? I feel like you are going to badly bias your selection here....

Comment (fioriNTU) [inline DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc]: Il 07/03/2016 16.54, Marco Musich ha scritto:
>
> In DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc 
> <https://github.com/cms-sw/cmssw/pull/13393#discussion_r55222938>:
>
> > -    }
> > -    auto subdetandlayer = findSubdetAndLayer(RawId, tTopo);
> > -    auto histos = m_SubdetLayerResiduals[subdetandlayer];
> > -    // fill if its error is not zero
> > -    if(it.resXprimeErr != 0 && histos.x.base) {
> > -      histos.x.base->Fill(it.resXprime);
> > -      histos.x.normed->Fill(it.resXprime/it.resXprimeErr);
> > -    }
> > -    if(it.resYprimeErr != 0 && histos.y.base) {
> > -      histos.y.base->Fill(it.resYprime);
> > -      histos.y.normed->Fill(it.resYprime/it.resYprimeErr);
> > +  avalidator_.fillTrackQuantities(iEvent, iSetup,
> > +                  // tell the validator to only look at good tracks
> > +                  [](const reco::Track& track) -> bool {
> > +                    return track.pt() > 0.75
> > +                        && abs( track.dxy() ) > 5*track.dxyError();
>
> @fioriNTU <https://github.com/fioriNTU> can you please comment? I feel 
> like you are going to badly bias your selection here....
>
There is an error in this cut, in the original code it was

if(abs( track.dxy() ) > 5*track.dxyError()) continue;

now the logic is inverted. Of course we don't want to select displaced tracks. Marcel update the PR, thanks for spotting!

> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/13393/files#r55222938>.
>



---
Questa e-mail è stata controllata per individuare virus con Avast antivirus.
https://www.avast.com/antivirus

Comment (mmusich) [inline DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc]: @fioriNTU  still if you cut on the dxy of the track the acceptance will vary with the phi of the track. IMHO you should cut on dxy(PV) or dxy(BS) to select good tracks. Well probably doesn't matter too much at this level...

Comment (fioriNTU) [inline DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc]: Yes Marco, that's right, even if I don't expect a big difference we will 
implement as you suggest

Il 07/03/2016 17.39, Marco Musich ha scritto:
>
> In DQM/TrackerMonitorTrack/src/MonitorTrackResiduals.cc 
> <https://github.com/cms-sw/cmssw/pull/13393#discussion_r55230869>:
>
> > -    }
> > -    auto subdetandlayer = findSubdetAndLayer(RawId, tTopo);
> > -    auto histos = m_SubdetLayerResiduals[subdetandlayer];
> > -    // fill if its error is not zero
> > -    if(it.resXprimeErr != 0 && histos.x.base) {
> > -      histos.x.base->Fill(it.resXprime);
> > -      histos.x.normed->Fill(it.resXprime/it.resXprimeErr);
> > -    }
> > -    if(it.resYprimeErr != 0 && histos.y.base) {
> > -      histos.y.base->Fill(it.resYprime);
> > -      histos.y.normed->Fill(it.resYprime/it.resYprimeErr);
> > +  avalidator_.fillTrackQuantities(iEvent, iSetup,
> > +                  // tell the validator to only look at good tracks
> > +                  [](const reco::Track& track) -> bool {
> > +                    return track.pt() > 0.75
> > +                        && abs( track.dxy() ) > 5*track.dxyError();
>
> @fioriNTU <https://github.com/fioriNTU> still if you cut on the dxy of 
> the track the acceptance will vary with the phi of the track. IMHO you 
> should cut on dxy(PV) or dxy(BS) to select good tracks. Well probably 
> doesn't matter too much at this level...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/13393/files#r55230869>.
>



---
Questa e-mail è stata controllata per individuare virus con Avast antivirus.
https://www.avast.com/antivirus

**Comment by schneiml** [comment]
The cut on dxy should now be corrected and also use the correct primary vertex. Now significantly more tracks pass the cut. Thanks @mmusich for paying attention...

**Comment by mmusich** [comment]
please test

**Comment by mmusich** [comment]
+1

**Comment by deguio** [comment]
+1

File: DQM/SiPixelCommon/python/SiPixelOfflineDQM_source_cff.py [hunk 3]
```diff
@@ -31,6 +31,10 @@
 from DQM.SiPixelMonitorTrack.SiPixelMonitorEfficiency_cfi import *
 SiPixelHitEfficiencySource.saveFile = False
 SiPixelHitEfficiencySource.trajectoryInput = cms.InputTag('generalTracks') 
+from DQM.TrackerMonitorTrack.SiPixelMonitorTrackResiduals_cfi import *
+SiPixelMonitorTrackResiduals.trajectoryInput = 'generalTracks'
+SiPixelMonitorTrackResiduals.Tracks          = 'generalTracks'
+SiPixelMonitorTrackResiduals.Mod_On          = False
```
Comment (davidlange6) [inline DQM/SiPixelCommon/python/SiPixelOfflineDQM_source_cff.py]: either modify the CFI with these changes or clone the module and give it a new name before changing its configuration
@schneiml @deguio

Comment (schneiml) [inline DQM/SiPixelCommon/python/SiPixelOfflineDQM_source_cff.py]: @davidlange6 what should the configuration in the _cfi do by default? As far as I see all usages of the old `MonitorTrackResiduals` override the default, so I don't know which scenario the default was good for...

Maybe the _cfi's should be changed to `generalTracks` (which seems to be most popular). Note also that some configs set `trajectoryInput` and `Tracks` to different values, which as far as I see, never made sense.

**Comment by davidlange6** [comment]
@schneiml - I guess I don't see SiPixelMonitorTrackResiduals used at all. maybe I just don't have the right search string.

**Comment by schneiml** [comment]
@davidlange6  `SiPixelMonitorTrackResiduals` is introduced in this PR, therefore this is also the only usage. 

However, `MonitorTrackResiduals` (the Strip version) was there before. The SiPixel cfi is a clone of the old cfi and I left in place the same values. 

Therefore, I think either both should be updated or both should stay as-is. In terms of compatibility just changing the SiPixel version is safer, but I see no reason at all to have the cfi's diverge, especially given that nobody seems to rely on the defaults anyways.

**Comment by davidlange6** [comment]
ah- but there is no need to clone a CFI to use it. You can either
1) make a new CFI with all of your changes
2) import the original module, clone it and change it

**Comment by schneiml** [comment]
Well, the single difference between the CFIs is the EDAnalyzer that they refer to... We have two very similar, but independent CMSSW plugins that handle SiStrip and SiPixel. It makes sense to have two very similar, but independent, CFI files that configure them... The original point was to have two plugins to keep SiPixel and SiStrip separated, see also @deguio 's comment: https://github.com/cms-sw/cmssw/pull/13393#issuecomment-193285278 .

I am not sure whether it is possible to change the plugin after a clone. If yes, that might be a very reasonable way to remove the duplication. But this was not entirely the original question. 

The original question is, for me at least, whether we should change the CFIs to have sane default values to avoid unnecessary options in config files, or whether I should assume that somebody has a good reason for the current defaults and they should stay as is to avoid breaking other peoples stuff (my overview of CMSSW is very limited).

**Comment by davidlange6** [comment]
i'm lost then, sorry. if the analyzers are different, then you certainly need a new cfi. but if you added a cfi, but the parameter values that you need inside. Its better yet to put sane values everywhere, but that is a broader question for tracker to agree on the approach for.

**Comment by venturia** [comment]
...and then document what is the agreed approach for future developers!

**Comment by venturia** [comment]
can you profit of this development to (start to ) write the documentation of this package in a README.md file which will then be displayed in github?

**Comment by schneiml** [comment]
@davidlange6 sorry to confuse you. The thing is that this is (to my knowledge) the first shared Strip/Pixel DQM plugin. So I had no reference to copy from. The approach that we chose (after a very hacky solution in #13240 ) is to generate two plugins from the same source code using template parameters such that the old separation is there again. We could do differently, but we did not. 

Now, what I need answered is whether to 1) change the new cfi 2) change both cfis or 3) merge as-is.

And @venturia , this is one of the few plugins that has a  README.md, and if you look into the diff, the new version even explains the template stuff and parameters.

**Comment by venturia** [comment]
right, I was looking at the wrong package. Thanks a lot!

**Comment by davidlange6** [comment]
Right - so I think to address my comment, the only change is to move
SiPixelMonitorTrackResiduals.trajectoryInput = 'generalTracks'
SiPixelMonitorTrackResiduals.Tracks          = 'generalTracks'
SiPixelMonitorTrackResiduals.Mod_On          = False

to your new SiPixelMonitorTrackResiduals_cfi.py

**Comment by mmusich** [comment]
please test

**Comment by schneiml** [comment]
@davidlange6 the I modified the CFI file now. I hope this actually works (let's see what the buildbot thinks -- my CMSSW checkout broke and I could only test after pushing).

**Comment by mmusich** [comment]
+1

**Comment by deguio** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**